src/cli/subcommands/car_cmd.rs

```diff
@@ -0,0 +1,236 @@
+                .unwrap();
+            car
+        }
+
+        fn into_stream(self) -> impl Stream<Item = BlockPair> {
+            futures::stream::iter(self.0.into_iter().map(|b| (b.cid, b.data)))
+        }
+
+        /// Implicit clone is performed inside to simplify caller code
+        fn to_stream(&self) -> impl Stream<Item = BlockPair> {
+            self.clone().into_stream()
+        }
+    }
+
+    impl Arbitrary for Blocks {
+        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+            // `CarReader` complains when n is 0: Error: Failed to parse CAR file: empty CAR file
+            let n = u16::arbitrary(g).saturating_add(1) as usize;
+            let mut blocks = Vec::with_capacity(n);
+            for _ in 0..n {
+                // use small len here to increase the chance of duplication
+                let data = [u8::arbitrary(g), u8::arbitrary(g)];
+                let cid = Cid::new_v1(DAG_CBOR, multihash::Code::Blake2b256.digest(&data));
+                let block = Block {
+                    cid,
+                    data: data.to_vec(),
+                };
+                blocks.push(block);
+            }
+            Self(blocks)
+        }
+    }
+
+    #[quickcheck]
+    fn blocks_roundtrip(blocks: Blocks) -> anyhow::Result<()> {
+        block_on(async move {
+            let car = blocks.into_car_bytes().await;
+            let mut reader = CarReader::new(car.as_slice()).await?;
+            let mut blocks2 = vec![];
+            while let Some(b) = reader.next_block().await? {
+                blocks2.push(b);
+            }
+            let blocks2 = Blocks(blocks2);
+            let car2 = blocks2.into_car_bytes().await;
+
+            assert_eq!(car, car2);
+
+            Ok::<_, anyhow::Error>(())
+        })
+    }
+
+    #[quickcheck]
+    fn dedup_block_stream_tests_a_a(a: Blocks) {
+        // ∀A. A∪A = A
+        assert_eq!(dedup_block_stream_wrapper(&a, &a), HashSet::from(&a));
+    }
+
+    #[quickcheck]
+    fn dedup_block_stream_tests_a_b(a: Blocks, b: Blocks) {
+        let union_a_b = dedup_block_stream_wrapper(&a, &b);
+        let union_b_a = dedup_block_stream_wrapper(&b, &a);
+        // ∀AB. A∪B = B∪A
+        assert_eq!(union_a_b, union_b_a);
+        // ∀AB. A⊆(A∪B)
+        union_a_b.is_superset(&HashSet::from(&a));
+        // ∀AB. B⊆(A∪B)
+        union_a_b.is_superset(&HashSet::from(&b));
+    }
+
+    fn dedup_block_stream_wrapper(a: &Blocks, b: &Blocks) -> HashSet<Cid> {
+        let blocks: Vec<Cid> =
+            block_on_stream(dedup_block_stream(a.to_stream().chain(b.to_stream())))
+                .map(|(cid, _)| cid)
+                .collect();
+
+        // Ensure `dedup_block_stream` works properly
+        assert!(blocks.iter().all_unique());
+
+        HashSet::from_iter(blocks)
+    }
+
+    #[quickcheck]
+    fn car_dedup_block_stream_tests(a: Blocks, b: Blocks) -> anyhow::Result<()> {
+        let cid_union: HashSet<Cid> = [a.0.as_slice(), b.0.as_slice()]
```

Comment:
Use `HashSet::from`

Reply:
Do you mean `HashSet::from_iter` or could you elabrate?